Add support for using declarations and async using (#672)
dotnet/csharpstandard: standard/statements.md

```csharp
using «local_variable_type» «local_variable_declarators»
    // statements
```

is equivalent to

```csharp
using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

and

```csharp
await using «local_variable_type» «local_variable_declarators»
    // statements
```

is equivalent to

```csharp
await using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

The lifetime of the variables declared in a *local_variable_declaration* extends to the end of the scope in which they are declared. Those variables are then disposed in the reverse order in which they are declared.

<!-- Example: {template:"code-in-partial-class", name:"LocalVariableDecls6", additionalFiles:["SupportLocalVarDecl.cs"], replaceEllipsis:true, customEllipsisReplacements: ["\"File1.txt\", FileMode.Create", "\"File2.txt\", FileMode.Create", "\"File3.txt\", FileMode.Create"]} -->
```csharp
static void M()
{
    using FileStream f1 = new FileStream(...);
    using FileStream f2 = new FileStream(...), f3 = new FileStream(...);
    ...
    // Dispose f3
    // Dispose f2 
    // Dispose f1
}
```

> A `goto` cannot jump to a location before a `using` declaration within the same block.

Reviewer: I think this is incorrect, the point is you cannot jump over using in either direction. Basing wording from the [LDM notes](https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2018/LDM-2018-12-03.md#goto-and-using-declarations) I’ll offer:

```suggestion
When jumping to a label in a statement list which contains *local_variable_declaration*s involving `using`, there must not be any such declaration between the `goto` statement and the label.```

Apart from wordsmithing this needs x-ref to goto, and goto probably needs this restriction listed in its description as well.

Also surely this is an echo of a similar statement for `using (…) {…}`, so is it needed here at all given the equivalency defined above – this is just a shorthand.

The compile-time processing of a `foreach` statement first determines the ***collection type*** (`C`), ***enumerator type*** (`E`) and ***iteration type*** (`T`, `ref T` or `ref readonly T`) of the expression.

The determination is similar for the synchronous and asynchronous versions. Different interfaces with different methods and return types distinguish the synchronous and asynchronous versions. The general process proceeds as follows. Names within '«' and '»' are placeholders for the actual names for synchronous and asynchronous iterators. The types allowed for «GetEnumerator», «MoveNext», «IEnumerable»\<T>, «IEnumerator»\<T>, and any other distinctions are detailed in [§13.9.5.2](statements.md#13952-synchronous-foreach) for a synchronous `foreach` statement, and in [§13.9.5.3](statements.md#13953-await-foreach) for an asynchronous `foreach` statement.

> 1. Determine whether the type `X` has an appropriate «GetEnumerator» method:

Reviewer 2: In this context, it's not clear where the type `X` comes from. Above there is a collection type `C` mentioned, is it the same type as that?

Reviewer 3: Good point, but I don't think it's `C`. The ***collection type*** (`C`) is determined on line 1126 by examining `X`.

Reviewer 3: The original text (_original_ line 1125) had ``If the type `X` of *expression*``. It was probably removed by mistake? @BillWagner

Reviewer 4: PR is here: https://github.com/dotnet/csharpstandard/pull/1422

Thanks for pointing it out @RikkiGibson

   1. Perform member lookup on the type `X` with identifier «GetEnumerator» and no type arguments. If the member lookup does not produce a match, or it produces an ambiguity, or produces a match that is not a method group, check for an enumerable interface as described in step 2. It is recommended that a warning be issued if member lookup produces anything except a method group or no match.
   1. Perform overload resolution using the resulting method group and an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, check for an enumerable interface as described below. It is recommended that a warning be issued if overload resolution produces anything except an unambiguous public instance method or no applicable methods.
   1. If the return type `E` of the «GetEnumerator» method is not a class, struct or interface type, produce an error and take no further steps.

>    1. Perform member on `E` with the identifier `Current` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a public instance property that permits reading, produce an error and take no further steps.

Reviewer 3: ```suggestion
   1. Perform member lookup on `E` with the identifier `Current` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a public instance property that permits reading, produce an error and take no further steps.
```

   1. Perform member lookup on `E` with the identifier «MoveNext» and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a method group, produce an error and take no further steps.
   1. Perform overload resolution on the method group with an empty argument list. If overload resolution results in: no applicable methods; an ambiguity; or a single best method but that method is either static, or not public, or its return type is not an allowed return type; then produce an error and take no further steps.
   1. The collection type is `X`, the enumerator type is `E`, and the iteration type is the type of the `Current` property.
1. Otherwise, check for an enumerable interface:
   1. If among all the types `Tᵢ` for which there is an implicit conversion from `X` to «IEnumerable»\<Tᵢ>, there is a unique type `T` such that `T` is not `dynamic` and for all the other `Tᵢ` there is an implicit conversion from «IEnumerable»\<T> to «IEnumerable»\<Tᵢ>, then the collection type is the interface «IEnumerable»\<T>, the enumerator type is the interface «IEnumerator»\<T>, and the iteration type is `T`.

>    1. Otherwise, if there is more than one such type `T`, then produce an error and take no further steps.

Reviewer 3: This just reminds me of the open bug about these lines: https://github.com/dotnet/csharpstandard/issues/1188.

> *Note*: If *expression* has the value `null`, a `System.NullReferenceException` is thrown at run-time. *end note*

An implementation is permitted to implement a given *foreach_statement* differently; e.g., for performance reasons, as long as the behavior is consistent with the expansions described in §13.9.5.2 and §13.9.5.3.

#### 13.9.5.2 Synchronous foreach

A synchronous `foreach` does not include the `await` keyword before the `foreach` keyword. The determination of ***collection type***, ***enumeration type*** and ***iteration type*** proceeds as described in §13.9.5.1, where:

- «GetEnumerator» is a `GetEnumerator` method.
- «MoveNext» is a `MoveNext` method with a `bool` return type.
- «IEnumerable»\<T> is the `System.Collections.Generic.IEnumerable<T>` interface.
- «IEnumerator»\<T> is the `System.Collections.Generic.IEnumerator<T>` interface.

In addition, the following modifications are made to the steps in §13.9.5.1:

Before the process described in §13.9.5.1, the following steps are taken:

> - If the type `X` of *expression* is an array type then there is an implicit reference conversion from `X` to the `IEnumerable<T>` interface where `T` is the is the element type of the array `X` (§17.2.3).

Reviewer 3: ```suggestion
- If the type `X` of *expression* is an array type then there is an implicit reference conversion from `X` to the `IEnumerable<T>` interface where `T` is the element type of the array `X` (§17.2.3).
```

- If the type `X` of *expression* is `dynamic` then there is an implicit conversion from *expression* to the `IEnumerable` interface ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)). The collection type is the `IEnumerable` interface and the enumerator type is the `IEnumerator` interface. If the `var` identifier is given as the *local_variable_type* then the iteration type is `dynamic`, otherwise it is `object`.

If the process in §13.9.5.1 completes without producing a single collection type, enumerator type, and iteration type, the following steps are taken:

- If there is an implicit conversion from `X` to the `System.Collections.IEnumerable` interface, then the collection type is this interface, the enumerator type is the interface `System.Collections.IEnumerator`, and the iteration type is `object`.
- Otherwise, an error is produced, and no further steps are taken.

> 
> A `foreach` statement of the form

Reviewer 3: ```suggestion
```

Reviewer 4: I think this is needed. Deferring to @RexJaeschke

Reviewer 3: This was my misunderstanding the diff, sorry!

The iteration variable corresponds to a local variable with a scope that extends over the embedded statement. During execution of a `foreach` statement, the iteration variable represents the collection element for which an iteration is currently being performed. If the iteration variable denotes a read-only variable, a compile-time error occurs if the embedded statement attempts to modify it (via assignment or the `++` and `--` operators) or pass it as a reference or output parameter.

The compile-time processing of a `foreach` statement first determines the ***collection type*** (`C`), ***enumerator type*** (`E`) and ***iteration type*** (`T`, `ref T` or `ref readonly T`) of the expression.

> The determination is similar for the synchronous and asynchronous versions. Different interfaces with different methods and return types distinguish the synchronous and asynchronous versions. The general process proceeds as follows:

Reviewer 5: I think we possibly need a sentence or two to explain what the identifiers within chevrons mean. (I get the drift, but we should probably be explicit.)

Reviewer 5: (Referring forward to 13.9.5.2 and 13.9.5.3 and saying those specify the actual names for each case would probably be sufficient.)

Reviewer 4: Added in the upcoming commit.

> The determination is similar for the synchronous and asynchronous versions. Different interfaces with different methods and return types distinguish the synchronous and asynchronous versions. The general process proceeds as follows. The text with '«' and '»' denote placeholder names for the method names where they are different for the synchronous and asynchronous iterators. Sections §13.9.5.2 and §13.9.5.3 provide the names for those placeholders in synchronous and asynchronous iterators, respectively.

Reviewer: "The text with «…” sounds wrong to me, maybe something along the lines of:

```suggestion
The determination is similar for the synchronous and asynchronous versions. Different interfaces with different methods and return types distinguish the synchronous and asynchronous versions. The general process proceeds as follows. Names within '«' and '»' are placeholders for the actual names for synchronous and asynchronous iterators; these are given in §13.9.5.2 and §13.9.5.3 for synchronous and asynchronous iterators respectively.
```

I think I’d prefer a table of substitutions here rather than later, but maybe that is just me!

- Determine whether the type `X` has an appropriate «GetEnumerator» method:

>   - Perform member lookup on the type `X` with identifier «GetEnumerator» and no type arguments. If the member lookup does not produce a match, or it produces an ambiguity, or produces a match that is not a method group, check for an enumerable interface as described below. It is recommended that a warning be issued if member lookup produces anything except a method group or no match.

Reviewer: I think using numbered rather than bulleted points could help here, so “check for an enumerable interface as described below” could be something like “continue at step 2” (the current step is 1.1).

Reviewer 4: tagging @jskeet I did this on the next commit. However, I don't know how the word converter will display nested numbered lists, and if that is consistent with any ECMA requirement. @RexJaeschke please add a comment if ECMA has a rule for this.

  - Perform overload resolution using the resulting method group and an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, check for an enumerable interface as described below. It is recommended that a warning be issued if overload resolution produces anything except an unambiguous public instance method or no applicable methods.
  - If the return type `E` of the «GetEnumerator» method is not a class, struct or interface type, an error is produced and no further steps are taken.

>   - Member lookup is performed on `E` with the identifier `Current` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a public instance property that permits reading, an error is produced and no further steps are taken.

Reviewer: The tense changes here, earlier bullets “Perform…” (a command), from here “… is performed” (descriptive); better to be consistent.

*(This error originated with some of the source material)*

  - Member lookup is performed on `E` with the identifier «MoveNext» and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a method group, an error is produced and no further steps are taken.

>   - Overload resolution is performed on the method group with an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, or its return type is not an allowed return type, an error is produced, and no further steps are taken., an error is produced, and no further steps are taken.

Reviewer: An editing snafu here, we all (well I do as well) suffer from these ;-( I’ll offer:

```suggestion
  - Overload resolution is performed on the method group with an empty argument list. If overload resolution results in: no applicable methods; an ambiguity; or a single best method but that method is either static, or not public, or its return type is not an allowed return type; then an error shall be produced and no further steps taken.
```

The tense may be wrong (shall be)?

  - The collection type is `X`, the enumerator type is `E`, and the iteration type is the type of the `Current` property.
- Otherwise, check for an enumerable interface:
  - If among all the types `Tᵢ` for which there is an implicit conversion from `X` to «IEnumerable»\<Tᵢ>, there is a unique type `T` such that `T` is not `dynamic` and for all the other `Tᵢ` there is an implicit conversion from «IEnumerable»\<T> to «IEnumerable»\<Tᵢ>, then the collection type is the interface «IEnumerable»\<T>, the enumerator type is the interface «IEnumerator»\<T>, and the iteration type is `T`.
  - Otherwise, if there is more than one such type `T`, then an error is produced and no further steps are taken.

> The types allowed for «GetEnumerator», «MoveNext», «IEnumerable»\<T>, «IEnumerator»\<T>, and any other distinctions are detailed in [§13.9.5.2](statements.md#13952-synchronous-foreach) for a synchronous `foreach` statement, and in [§13.9.5.3](statements.md#13953-await-foreach) for an asynchronous `foreach` statement.

Reviewer: This is a small variation of the text before the above steps, better to combine both into the one above and drop this one.

> *Note*: If *expression* has the value `null`, a `System.NullReferenceException` is thrown at run-time. *end note*

> An implementation is permitted to implement a given *foreach_statement* differently; e.g., for performance reasons, as long as the behavior is consistent with this expansion.

Reviewer: This is now out of place, the expansions come in the following clauses

#### 13.9.5.2 Synchronous foreach

A synchronous `foreach` does not include the `await` keyword before the `foreach` keyword. The determination of ***collection type***, ***enumeration type*** and ***iteration type*** proceeds as described in §13.9.5.1, where:

- «GetEnumerator» is a `GetEnumerator` method.
- «MoveNext» is a `MoveNext` method with a `bool` return type.
- «IEnumerable»\<T> is the `System.Collections.Generic.IEnumerable<T>` interface.
- «IEnumerator»\<T> is the `System.Collections.Generic.IEnumerator<T>` interface.

In addition, the following modifications are made to the steps in §13.9.5.1:

Before the process described in §13.9.5.1, the following steps are taken:

> - If the type `X` of *expression* is an array type then there is an implicit reference conversion from X to the `IEnumerable` interface (since `System.Array` implements this interface). The collection type is the `IEnumerable` interface, the enumerator type is the `IEnumerator` interface and the iteration type is the element type of the array type `X`. (sync)

Reviewer 5: ```suggestion
- If the type `X` of *expression* is an array type then there is an implicit reference conversion from `X` to the `IEnumerable` interface (since `System.Array` implements this interface). The collection type is the `IEnumerable` interface, the enumerator type is the `IEnumerator` interface and the iteration type is the element type of the array type `X`. (sync)
```

Reviewer 5: Are the "sync" and "sync only" bits in brackets still needed? This is already in a sync-only section. I suspect they're vestigial from when we were trying to unify things.

> - If the type `X` of *expression* is an array type then there is an implicit reference conversion from `X` to the `IEnumerable` interface (since `System.Array` implements this interface). The collection type is the `IEnumerable` interface, the enumerator type is the `IEnumerator` interface and the iteration type is the element type of the array type `X`. (sync)

Reviewer 5: In 17.2.3 we already say:

> A single-dimensional array `T[]` implements the interface `System.Collections.Generic.IList<T>` and its base interfaces.

So can't we say the collection type is `IEnumerable<A>` and the interface type is `IEnumerator<A>` where `A` is the element type of the array type `X`?

Reviewer 4: Updated per comment.

- If the type `X` of *expression* is `dynamic` then there is an implicit conversion from *expression* to the `IEnumerable` interface ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)). The collection type is the `IEnumerable` interface and the enumerator type is the `IEnumerator` interface. If the `var` identifier is given as the *local_variable_type* then the iteration type is `dynamic`, otherwise it is `object`. (sync only)

> If the process in §13.9.5.1 completes without produce a single collection type, enumerator type, and iteration type, the following steps are taken:

Reviewer 5: ```suggestion
If the process in §13.9.5.1 completes without producing a single collection type, enumerator type, and iteration type, the following steps are taken:
```

> - Otherwise, if there is an implicit conversion from `X` to the `System.Collections.IEnumerable` interface, then the collection type is this interface, the enumerator type is the interface `System.Collections.IEnumerator`, and the iteration type is `object`. (sync only)

Reviewer 5: ```suggestion
- If there is an implicit conversion from `X` to the `System.Collections.IEnumerable` interface, then the collection type is this interface, the enumerator type is the interface `System.Collections.IEnumerator`, and the iteration type is `object`.
```

Reviewer 5: ("Otherwise" at the start of a list is weird, and this is already in a sync-only section.)

#### 13.9.5.3 await foreach

An asynchronous foreach uses the `await foreach` syntax. The determination of ***collection type***, ***enumeration type*** and ***iteration type*** proceeds as described in §13.9.5.1, where:

> - «GetEnumerator» is a `GetEnumeratorAsync` method.

Reviewer 3: ```suggestion
- «GetEnumerator» is a `GetEnumeratorAsync` method that has an awaitable return type (§12.9.8.2).
```

> - «MoveNext» is a `MoveNextAsync` method has an awaitable return type ([§12.9.8.2](expressions.md#12982-awaitable-expressions)) where the *await_expression* is classified as a `bool` ([§12.9.8.3](expressions.md#12983-classification-of-await-expressions)).

Reviewer 3: ```suggestion
- «MoveNext» is a `MoveNextAsync` method that has an awaitable return type ([§12.9.8.2](expressions.md#12982-awaitable-expressions)) where the *await_expression* is classified as a `bool` ([§12.9.8.3](expressions.md#12983-classification-of-await-expressions)).
```

  - Otherwise, if there is more than one such type `T`, then an error is produced and no further steps are taken.
  - Otherwise, an error is produced, and no further steps are taken.
It is an error for the ***iteration type*** of an `await foreach` statement to be a reference variable (§9.7).

The above steps, if successful, unambiguously produce a collection type `C`, enumerator type `E` and iteration type `T`. An `await foreach` statement of the form

```csharp
await foreach (V v in x) «embedded_statement»
```

is then equivalent to:

```csharp
{
    E e = ((C)(x)).GetAsyncEnumerator();
    try
    {
        while (await e.MoveNextAsync())
        {
            V v = (V)(T)e.Current;
            «embedded_statement»
        }
    }
    finally
    {
        ... // Dispose e
    }
}

> ```

Reviewer 3: I think we need the expansion to be defined for the async case too. It looks like we're losing that.

The async case is distinct from the expansion shown under "synchronous foreach" because it involves an `await` and then has some interesting rules for how disposal works.

Observations from my testing:
- `await using` will not dispose the enumerator in a synchronous fashion regardless of whether it implements `IDisposable` or has a `public void Dispose()` method. (edit: true, but `await foreach` is the same)
- ~~`await using`~~ `await foreach` will call `DisposeAsync()` whether or not the enumerator implements the `IAsyncDisposable` interface, so long as the return type is awaitable. (The awaited result type doesn't even have to be void!) It also doesn't have to be public, so long as it's accessible at the `await foreach` location. If accessible, it is preferred over an implicit implementation of `IAsyncDisposable.DisposeAsync`.
  - In all of these points, it differs from the synchronous version.
  - This means that `await foreach` disposes entirely differently from how `await using` disposes.

~~Not sure how much of that is spec material vs compiler quirks.~~ or me setting up my tests incorrectly 😅

Reviewer 4: See commit 2a7718554e96ae52a143b576c3370778b212f4fd, the 37th in this PR.

Reviewer 3: Oh! That is a good addition. What I see _here_ though that I still think is lacking that was being deleted by this PR was the expansion for `await foreach`.

Reviewer 4: It was added in an earlier PR. See lines 1353-1368

Reviewer 3: Ah, thank you, I was misunderstanding diffs again.

Reviewer 3: I had erroneously concluded that `await foreach` had a fallback to a synchronous dispose, but that was my mistake in testing. The rules are as specified at <https://github.com/dotnet/csharplang/blob/main/meetings/2019/LDM-2019-01-16.md#pattern-based-disposal-in-await-foreach> which @BillWagner found.

#### 13.9.5.3 await foreach

An asynchronous foreach uses the `await foreach` syntax. The determination of ***collection type***, ***enumeration type*** and ***iteration type*** proceeds as described in §13.9.5.1, where:

- «GetEnumerator» is a `GetEnumeratorAsync` method.
- «MoveNext» is a `MoveNextAsync` method has an awaitable return type ([§12.9.8.2](expressions.md#12982-awaitable-expressions)) where the *await_expression* is classified as a `bool` ([§12.9.8.3](expressions.md#12983-classification-of-await-expressions)).
- «IEnumerable»\<T> is the `System.Collections.Generic.IAsyncEnumerable<T>` interface.
- «IEnumerator»\<T> is the `System.Collections.Generic.IAsyncEnumerator<T>` interface.

> It is an error for the ***iteration type*** of an `await foreach` to be a reference variable (§9.7).

Reviewer 5: ```suggestion
It is an error for the ***iteration type*** of an `await foreach` statement to be a reference variable (§9.7).
```

(This feels better to me, and is consistent with the next line. It's not consistent with the first line of the section, admittedly.)

> An `await foreach` statement of the form

Reviewer 5: In the equivalent code below, a cancellation token is never passed to MoveNextAsync. I can't remember the details (is this where the attribute comes in - is it passed in automatically?) but I'm pretty sure a change is needed here.

Reviewer 4: I won't get to it before the meeting, but I'll add a second code sample following the meeting.

Reviewer 4: I'd had this wrong. The cancellation token (or combined token) is passed to `GetAsyncEnumerator`. Adding a second sample that includes the `WithCancellation` extension method.

Reviewer 3: A combined token isn't passed to GetAsyncEnumerator, but rather a second token is passed to it, and GetAsyncEnumerator then performs the combining between that token and the token that was passed to the async iterator method.

}
finally
{
    // enumerator.Dispose() if enumerator is a `ref struct` type and supports synchronous disposal 
    await enumerator.DisposeAsync(); 
    // omitted, along with the try/finally,

>     // if the enumerator doesn't expose a dispose method

Reviewer: These four lines are confusing and doesn’t fully cover the cases in the explanation just below starting on line 1372. However the explanation doesn’t include the statement here that the try/finally can be completely omitted, but it does have a case where the finally is empty. These discrepancies need fixing.

Maybe the code sample could just include a comment referring to line 1372 for the expansion, and the explanation be expanded to cover the lack of a dispose method producing an empty finally?

Such a change would remove the comment here about eliding the try/finally and replace it with a statement in the explanation that the finally is empty for that case (note there is an existing empty case, the conditions for each could be combined?). I can’t see this is an issue, is not removing an empty finally a valid implementation optimisation?

Reviewer 4: Good catch. Coming in next commit.

}
finally
{

>     // enumerator.Dispose() if enumerator is a `ref struct` type and supports synchronous disposal 

Reviewer 5: Leaving a comment here as the closest change - line 1345 needs to be removed.

    await enumerator.DisposeAsync(); 

>     // omitted, along with the try/finally,

Reviewer 5: The order of these comments is confusing me. I think it would be worth rewriting these two slightly as "The try/finally is omitted entirely (including the finally body) if the enumerator doesn't expose a dispose method" and *then* have the method calls.

But it may just be me - see what other people think.

Reviewer 4: The combination of your comments made me do a bit of spec refactoring. I think it reads better and is simpler now.

The body of the `finally` block is constructed according to the following steps:

- If `E` has an accessible `DisposeAsync()` method where the return type is awaitable (§12.9.8.2), the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        await e.DisposeAsync();
    }
    ```

- Otherwise, if there is an implicit conversion from `E` to the `System.IAsyncDisposable` interface, then
  - If `E` is a non-nullable value type then the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        await ((System.IAsyncDisposable)e).DisposeAsync();
    }
    ```

  - Otherwise the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {

>         System.IDisposable d = e as System.IAsyncDisposable;

Reviewer 3: ```suggestion
        System.IAsyncDisposable d = e as System.IAsyncDisposable;
```

        if (d != null)
        {
            await d.DisposeAsync();
        }
    }
    ```

>     except that if `E` is a value type, or a type parameter instantiated to a value type, then the conversion of `e` to `System.IAsyncDisposable` shall not cause boxing to occur.

Reviewer 3: ~~Since it may be System.IDisposable as well:~~

Reviewer 3: Also, should this message appear underneath the final expansion showing an interface cast?

Reviewer 3: I had erroneously concluded that `await foreach` had a fallback to a synchronous dispose, but that was my mistake in testing. The rules are as specified at <https://github.com/dotnet/csharplang/blob/main/meetings/2019/LDM-2019-01-16.md#pattern-based-disposal-in-await-foreach> which @BillWagner found.

Reviewer 5: So does the bullet below need to be removed entirely? I'm confused now...

Reviewer 3: I think so. A ref struct cannot be used as the enumerator of `await foreach` because a ref struct cannot be stored on the heap to enable access to it across an `await` or `yield` boundary.

}
```

> In the case where the expression `enumerable` represents a method call expression and one of the parameters is marked with the `EnumeratorCancellationAttribute` (§enumerator-cancellation) the a `CancellationToken` is passed to the `GetAsyncEnumerator` method. Other library methods may require a `CancellationToken` is passed to `GetAsyncEnumerator`. When those methods are part of the expression `enumerable`, the tokens shall be combined into a single token as if by `CreateLinkedTokenSource` and its `Token` property.

Reviewer 5: ```suggestion
In the case where the expression `enumerable` represents a method call expression and one of the parameters is marked with the `EnumeratorCancellationAttribute` (§enumerator-cancellation) the `CancellationToken` is passed to the `GetAsyncEnumerator` method. Other library methods may require a `CancellationToken` is passed to `GetAsyncEnumerator`. When those methods are part of the expression `enumerable`, the tokens shall be combined into a single token as if by `CreateLinkedTokenSource` and its `Token` property.
```

> In the case where the expression `enumerable` represents a method call expression and one of the parameters is marked with the `EnumeratorCancellationAttribute` (§enumerator-cancellation) the a `CancellationToken` is passed to the `GetAsyncEnumerator` method. Other library methods may require a `CancellationToken` is passed to `GetAsyncEnumerator`. When those methods are part of the expression `enumerable`, the tokens shall be combined into a single token as if by `CreateLinkedTokenSource` and its `Token` property.

Reviewer 5: Nit: `enumerable` could be a method call expression, but not specify an argument for a cancellation token due to the parameter being optional. I'd suggest that if we agree that *is* a problem, we address it in a separate PR.

The body of the `finally` block is constructed according to the following steps:

- If `E` has an accessible `DisposeAsync()` method where the return type is awaitable (§12.9.8.2), the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        await e.DisposeAsync();
    }
    ```

- Otherwise, if there is an implicit conversion from `E` to the `System.IAsyncDisposable` interface, then
  - If `E` is a non-nullable value type then the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        await ((System.IAsyncDisposable)e).DisposeAsync();
    }
    ```

  - Otherwise the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        System.IDisposable d = e as System.IAsyncDisposable;
        if (d != null)
        {
            await d.DisposeAsync();
        }
    }
    ```

    except that if `E` is a value type, or a type parameter instantiated to a value type, then the conversion of `e` to `System.IAsyncDisposable` shall not cause boxing to occur.

> - Otherwise, if `E` is a sealed type, the `finally` clause is expanded to an empty block:

Reviewer 5: Assuming I've understood the meeting properly, we need another bullet point about IDisposable

Reviewer 4: Added in final commit.

}
```

> In the case where the expression `enumerable` represents a method call expression and one of the parameters is marked with the `EnumeratorCancellationAttribute` (§enumerator-cancellation) the a `CancellationToken` is passed to the `GetAsyncEnumerator` method. Other library methods may require a `CancellationToken` is passed to `GetAsyncEnumerator`. When those methods are part of the expression `enumerable`, the tokens shall be combined into a single token as if by `CreateLinkedTokenSource` and its `Token` property.

Reviewer 3: I'm very happy with this language. Thank you @BillWagner for the large amount of effort you put into this.

}
```

> TODO:  Add second example that includes `.WithCancellation`

Reviewer 5: Adding a comment here to highlight that we still have a TODO.

Reviewer 4: My notes indicate that we aren't adding this expansion.

}
```

> In either expansion, an implementation may determine that a `CancellationToken` should be passed to the `GetAsyncEnumerator` method. This may include a combined token as shown in the example in §enumerator-cancellation. This enables the type implementing `IAsyncEnumerator` to implement cancellation as part of the implementation of `MoveNextAsync`.

Reviewer 4: Note for @jskeet and @Nigel-Ecma 

This still feels a bit hand-wavy to me. Any suggestions are welcome.

Reviewer: If the “In either expansion” is referring to the sync & async versions then the names (`GetAsyncEnumerator` *et al*)in this para are wrong.

It is certainly “hand-wavy” with “an implementation may determine”, TODO – come back here before submitting review!

    ```

    except that if `E` is a value type, or a type parameter instantiated to a value type, then the conversion of `e` to `System.IAsyncDisposable` shall not cause boxing to occur.
- Otherwise, if `E` is a `ref struct` type and has an accessible `Dispose()` method, the `finally` clause is expanded to the semantic equivalent of:

    ```csharp
    finally
    {
        e.Dispose();
    }
    ```

>   - Otherwise the `finally` clause is expanded to the semantic equivalent of:

Reviewer 5: This otherwise is nested, but I'm unclear why. If it's meant to be "E is a ref struct type without an accessible Dispose() method" then I think we should restructure the preceding part to be a peer, e.g.

```
- Otherwise, if `E` is a `ref struct` type:
  - If `E` has an accessible `Dispose()` method, the `finally` clause is expanded [...]
  - Otherwise, the `finally` clause is expanded to the semantic equivalent of...
```

Local variables declared in a *resource_acquisition* are read-only, and shall include an initializer. A compile-time error occurs if the embedded statement attempts to modify these local variables (via assignment or the `++` and `--` operators), take the address of them, or pass them as `ref` or `out` parameters.

A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposaAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.

A `using` statement of the form

```csharp
using (ResourceType resource = «expression» ) «statement»
```

> corresponds to one of three possible expansions. For class and non-ref struct resources, when `ResourceType` is a non-nullable value type or a type parameter with the value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints)), the expansion is semantically equivalent to

Reviewer 4: Copied from [#606 (comment)](https://github.com/dotnet/csharpstandard/pull/606/files#r940394361)

> Needs similar lowering rules for await using, particularly that the result of DisposeAsync is awaited rather than discarded.

Reviewer 4: Added in latest commit.

    ;
```

A ***resource*** is either a class or non-ref struct that implements either or both of the `System.IDisposable` or `System.`IAsync`Disposable` interfaces, which includes a single parameterless method named `Dispose` and/or `DisposeAsync`; or a ref struct that includes a method named `Dispose` having the same signature as that declared by `System.IDisposable`. Code that is using a resource can call `Dispose` or `DisposeAsync` to indicate that the resource is no longer needed.

If the form of *resource_acquisition* is *local_variable_declaration* then the type of the *local_variable_declaration* shall be either dynamic or a resource type. If the form of *resource_acquisition* is *expression* then this expression shall have a resource type. If `await` is present, the resource type shall implement `System.IAsyncDisposable`.

Local variables declared in a *resource_acquisition* are read-only, and shall include an initializer. A compile-time error occurs if the embedded statement attempts to modify these local variables (via assignment or the `++` and `--` operators), take the address of them, or pass them as reference or output parameters.

> A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposaAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.

Reviewer 3: ```suggestion
A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposeAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.
```

## 13.14 The using statement

### §using-general General

The `using` statement obtains one or more resources, executes a statement, and then disposes of the resource.

```ANTLR
using_statement
    : 'await'? 'using' '(' resource_acquisition ')' embedded_statement
    ;

resource_acquisition
    : non_ref_local_variable_declaration
    | expression
    ;

non_ref_local_variable_declaration
    : implicitly_typed_local_variable_declaration
    | explicitly_typed_local_variable_declaration
    ;
```

> A ***resource*** is either a class or non-ref struct that implements either or both of the `System.IDisposable` or `System.`IAsync`Disposable` interfaces, which includes a single parameterless method named `Dispose` and/or `DisposeAsync`; or a ref struct that includes a method named `Dispose` having the same signature as that declared by `System.IDisposable`. Code that is using a resource can call `Dispose` or `DisposeAsync` to indicate that the resource is no longer needed.

Reviewer 5: ```suggestion
A ***resource*** is either a class or non-ref struct that implements either or both of the `System.IDisposable` or `System.IAsyncDisposable` interfaces, which includes a single parameterless method named `Dispose` and/or `DisposeAsync`; or a ref struct that includes a method named `Dispose` having the same signature as that declared by `System.IDisposable`. Code that is using a resource can call `Dispose` or `DisposeAsync` to indicate that the resource is no longer needed.
```

If the form of *resource_acquisition* is *local_variable_declaration* then the type of the *local_variable_declaration* shall be either `dynamic` or a resource type. If the form of *resource_acquisition* is *expression* then this expression shall have a resource type. If `await` is present, the resource type shall implement `System.IAsyncDisposable`.

> *Note:* A `ref struct` type cannot be the resource type for a `using` statement with the `await` modifier because `ref struct` types cannot implement interfaces. *end note*

Local variables declared in a *resource_acquisition* are read-only, and shall include an initializer. A compile-time error occurs if the embedded statement attempts to modify these local variables (via assignment or the `++` and `--` operators), take the address of them, or pass them as reference or output parameters.

> A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposaAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.

Reviewer 5: ```suggestion
A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposeAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.
```

## 13.14 The using statement

### §using-general General

The `using` statement obtains one or more resources, executes a statement, and then disposes of the resource.

```ANTLR
using_statement
    : 'await'? 'using' '(' resource_acquisition ')' embedded_statement
    ;

resource_acquisition
    : non_ref_local_variable_declaration
    | expression
    ;

non_ref_local_variable_declaration
    : implicitly_typed_local_variable_declaration
    | explicitly_typed_local_variable_declaration
    ;
```

> A ***resource*** is either a class or non-ref struct that implements either or both of the `System.IDisposable` or `System.IAsyncDisposable` interfaces, which includes a single parameterless method named `Dispose` and/or `DisposeAsync`; or a ref struct that includes a method named `Dispose` having the same signature as that declared by `System.IDisposable`. Code that is using a resource can call `Dispose` or `DisposeAsync` to indicate that the resource is no longer needed.

Reviewer: The terms *resource* and *resource type* seem to be getting used somewhat interchangeably in this PR to mean the latter – a type, and *resource* is also being use to mean the acquired resource – a thing. The definition here of *resource* defines it as a type, not a thing. It can be a bit confusing.

If the form of *resource_acquisition* is *local_variable_declaration* then the type of the *local_variable_declaration* shall be either `dynamic` or a resource type. If the form of *resource_acquisition* is *expression* then this expression shall have a resource type. If `await` is present, the resource type shall implement `System.IAsyncDisposable`.

> > *Note:* A `ref struct` type cannot be the resource type for a `using` statement with the `await` modifier because `ref struct` types cannot implement interfaces. *end note*

Reviewer: This seems an odd note to me. In the synchronous case we have “a ref struct that includes a method named `Dispose` having the same signature as that declared by `System.IDisposable`” – the inability to implement an interface is no impediment in this case, so stating it is in the asynchronous case just seems, I’ll say “odd”.

Is this more a specific implementation limitation that should not be Standardised? Or is there a semantic reason for the limitation, and if so should that not be documented?

Reviewer 4: Edited. I made this normative, and removed the "because..." clause. The grammar would allow `await using ...` with a `ref struct`, but the earlier paragraphs note that `DisposeAsync` must be a member of `IAsyncDisposable` (not pattern based.)

Local variables declared in a *resource_acquisition* are read-only, and shall include an initializer. A compile-time error occurs if the embedded statement attempts to modify these local variables (via assignment or the `++` and `--` operators), take the address of them, or pass them as reference or output parameters.

> A `using` statement is translated into three parts: acquisition, usage, and disposal. Usage of the resource is implicitly enclosed in a `try` statement that includes a `finally` clause. This `finally` clause disposes of the resource. If a `null` resource is acquired, then no call to `Dispose` (or `DisposeAsync`) is made, and no exception is thrown. If the resource is of type `dynamic` it is dynamically converted through an implicit dynamic conversion ([§10.2.10](conversions.md#10210-implicit-dynamic-conversions)) to `IDisposable` (or `IAsyncDisposable`) during acquisition in order to ensure that the conversion is successful before the usage and disposal.

Reviewer 5: Not sure about "a null resource" - perhaps "if the result of the acquisition expression is a null reference"? (Further wordsmithing required, but I didn't want to spend too much time on it if everyone else is happy as it's already in the PR...)

Reviewer 4: Updated and a bit of wordsmithing.

A `using` statement of the form

```csharp
using (ResourceType resource = «expression» ) «statement»
```

> corresponds to one of three possible expansions. For class and non-ref struct resources, when `ResourceType` is a non-nullable value type or a type parameter with the value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints)), the expansion is semantically equivalent to

Reviewer: *(I’ve come back here, earlier comments below still use “expansion” – read them in the light of this comment)*

These are not “expansions” *per se* if the expression is “semantically equivalent to”. But what do we call them? *Meanings* maybe? If that is chosen we would here have:

```suggestion
has one of three possible meanings. For class and non-ref struct resources, when `ResourceType` is a non-nullable value type or a type parameter with the value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints)), the statement is semantically equivalent to
```

and similar changes on the following lines.

There might be something better than “meanings”, but they are not “expansions” as they can be implemented differently.

Reviewer 4: With the help of a thesaurus, I chose "formulation". I welcome alternatives. 

Note: I rejected "form" because we use it in a related way in many places.

Reviewer 3: The spec uses 'expansions' ~19 times, for instance to define the `as` operator (https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#121213-the-as-operator):

> If the compile-time type of `E` is not `dynamic`, the operation `E as T` produces the same result as
> 
> ```csharp
> E is T ? (T)(E) : (T)null
> ```
> 
> except that `E` is only evaluated once. A compiler can be expected to optimize `E as T` to perform at most one runtime type check as opposed to the two runtime type checks implied by the expansion above.
> 
> If the compile-time type of `E` is `dynamic`, unlike the cast operator the `as` operator is not dynamically bound ([§12.3.3](expressions.md#1233-dynamic-binding)). Therefore the expansion in this case is:
> 
> ```csharp
> E is T ? (T)(object)(E) : (T)null
> ```

Are these not examples of spec using "expansion" to describe semantic equivalence?

Depending on the outcome here: foreach statement 'expansions' are discussed in some of the edited text in this file.

Reviewer 4: Decision in 9/3 meeting: We agree that "semantic equivalent" is better. Captured in #1409

}
```

> For ref struct resources, the only possible expansion is

Reviewer: The first expansion “is semantically equivalent to” (1926), the second (1944) and this one are just “is” and then on 2002 we have “An implementation is permitted to implement a given *using_statement* differently” – which is the point of saying “semantically equivalent to”.

Repeating “semantically equivalent to” on 1944 & here, and dropping 2002, would be more consistent – if desired a *Note* could replace 2002 to emphasise the purpose of “semantically equivalent to” and even include the example suggestion of why an implementation might do it differently.

```csharp
{

>     ResourceType resource = «expression»;

Reviewer 5: I wonder whether `ResourceType` ought to be in chevrons here and above? (It's a bit of a third case - it's not "something from the original code" nor is it "fixed text"...)

Reviewer 4: updated.

    try
    {
        «statement»;
    }
    finally
    {
        resource.Dispose();
    }
}
```

In any expansion, the `resource` variable is read-only in the embedded statement, and the `d` variable is inaccessible in, and invisible to, the embedded statement.

> An implementation is permitted to implement a given *using_statement* differently, e.g., for performance reasons, as long as the behavior is consistent with the above expansion.

Reviewer: See comment on 1984 above

Reviewer 4: Deleted this paragraph.

using («expression») «statement»
```

> has the same possible expansions. In this case `ResourceType` is implicitly the compile-time type of the *expression*, if it has one. Otherwise, the interface `IDisposable` itself is used as the `ResourceType`. The `resource` variable is inaccessible in, and invisible to, the embedded *statement*.

Reviewer 5: Surely the expression *must* have a compile-time type, otherwise it can't conform with the earlier requirement of:

> If the form of *resource_acquisition* is *expression* then this expression shall have a resource type. If `await` is present, the resource type shall implement `System.IAsyncDisposable`.

Reviewer 4: updated in next commit.

using («expression») «statement»
```

> has the same possible expansions. `ResourceType` is implicitly the compile-time type of the *expression*, which must be a resource type.

Reviewer: This is echoing the statement on 1912, not sure if this repetition is good or bad…

Also see the comment on “expansions” above (1926)

Reviewer 4: I'm all for removing duplication.

```csharp
using «local_variable_type» «local_variable_declarators»
// statements
```

is semantically equivalent to

```csharp
using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

and

```csharp
await using «local_variable_type» «local_variable_declarators»
// statements
```

is semantically equivalent to

```csharp
await using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

The lifetime of the variables declared in a *non_ref_local_variable_declaration* extends to the end of the scope in which they are declared. Those variables are then disposed in the reverse order in which they are declared.

<!-- Example: {template:"code-in-partial-class", name:"LocalVariableDecls6", additionalFiles:["SupportLocalVarDecl.cs"], replaceEllipsis:true, customEllipsisReplacements: ["\"File1.txt\", FileMode.Create", "\"File2.txt\", FileMode.Create", "\"File3.txt\", FileMode.Create"]} -->
```csharp
static void M()
{
    using FileStream f1 = new FileStream(...);
    using FileStream f2 = new FileStream(...), f3 = new FileStream(...);
    ...
    // Dispose f3
    // Dispose f2 
    // Dispose f1
}
```

> A using declaration shall not appear directly inside a `case` label, but, instead, may be within a block inside a `case` label.

Reviewer 3: To cover `default` labels as well:
```suggestion
A using declaration shall not appear directly inside a *switch_label*, but, instead, may be within a block inside a *switch_label*.
```

>
> *end example*

A `using` statement of the form

```csharp
await using (ResourceType resource = «expression» ) «statement»
```

corresponds to the expansions shown above with `IAsyncDisposable` instead of `IDisposable`, `DisposeAsync` instead of `Dispose`, and the `Task` returned from `DisposeAsync` is `await`ed. An `await using` of the form:

> ```csharp

Reviewer 5: This duplication feels odd to me. Can we not just amend the previous paragraph to end with "The corresponding expansion is"

await using (ResourceType resource = «expression» ) «statement»
```

corresponds to

```csharp
{
    ResourceType resource = «expression»;
    try
    {
        «statement»;
    }
    finally
    {
        IAsyncDisposable d = (IAsyncDisposable)resource;
        if (d != null)
        {
            await d.DisposeAsync();
        }
    }
}
```

> When `ResourceType` is a reference type that implements `IAsyncDisposable`. Other expansions for `await using` perform similar substitutions from the synchronous `Dispose` method to the asynchronous `DisposeAsync` method.

Reviewer 5: If this is meant to be a condition of the expansion, I think it would definitely be better *before* the expansion. (At the moment it's written as if it's a standalone sentence, but it doesn't really work like that.)

Reviewer 4: Moved above the expansions.

> *Note*: Any jump statements (§13.10) in the *embedded_statement* must conform to expanded form of the `using` statement. *end note*

### §using-declarations Using declaration

A syntactic variant of the using statement is a *using declaration*.

```ANTLR
using_declaration
    : 'await'? 'using' non_ref_local_variable_declaration ';' statement_list?
    ;
```

A *using declaration* has the same semantics as, and can be rewritten as, the corresponding resource-acquisition form of the using statement (§using-general), as follows:

```csharp
using «local_variable_type» «local_variable_declarators»
// statements
```

is equivalent to

```csharp
using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

and

```csharp
await using «local_variable_type» «local_variable_declarators»
// statements
```

is equivalent to

```csharp
await using («local_variable_type» «local_variable_declarators»)
{
    // statements
}
```

> The lifetime of the variables declared in a *local_variable_declaration* extends to the end of the scope in which they are declared. Those variables are then disposed in the reverse order in which they are declared.

Reviewer 5: Should *local_variable_declaration* here be *non_ref_local_variable_declaration*?

>
> *end example*

When `ResourceType` is a reference type that implements `IAsyncDisposable`. Other expansions for `await using` perform similar substitutions from the synchronous `Dispose` method to the asynchronous `DisposeAsync` method. An `await using` statement of the form

```csharp
await using (ResourceType resource = «expression» ) «statement»
```

> corresponds to the expansions shown below with `IAsyncDisposable` instead of `IDisposable`, `DisposeAsync` instead of `Dispose`, and the `Task` returned from `DisposeAsync` is `await`ed:

Reviewer: “corresponds to” -> semantically equivalent to? And below.

Reviewer 4: semantically equivalent.

```csharp
await using (ResourceType resource = «expression» ) «statement»
```

corresponds to

```csharp
{
    ResourceType resource = «expression»;
    try
    {
        «statement»;
    }
    finally
    {
        IAsyncDisposable d = (IAsyncDisposable)resource;
        if (d != null)
        {
            await d.DisposeAsync();
        }
    }
}
```

> *Note*: Any jump statements (§13.10) in the *embedded_statement* must conform to expanded form of the `using` statement. *end note*

### §using-declarations Using declaration

A syntactic variant of the using statement is a *using declaration*.

```ANTLR
using_declaration
    : 'await'? 'using' non_ref_local_variable_declaration ';' statement_list?
    ;
```

A *using declaration* has the same semantics as, and can be rewritten as, the corresponding resource-acquisition form of the using statement (§using-general), as follows:

```csharp
using «local_variable_type» «local_variable_declarators»
// statements
```

> is equivalent to

Reviewer: (a self-critique – I know I am the source of this and 2124 :-))

Given other uses around here maybe “is semantically equivalent to” or “is equivalent in meaning to” or something similar?

Reviewer 4: "semantically equivalent" was already used often in the standard. I chose that option.
